src/internal/dbutil/tx.go

```diff
@@ -0,0 +1,97 @@
+package dbutil
+
+import (
+	"context"
+	"database/sql"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	"github.com/lib/pq"
+	"github.com/pachyderm/pachyderm/v2/src/internal/backoff"
+	"github.com/sirupsen/logrus"
+)
+
+type withTxConfig struct {
+	sql.TxOptions
+	MaxRetries int
+}
+
+// WithTxOption parameterizes the WithTx function
+type WithTxOption func(c *withTxConfig)
+
+// WithIsolationLevel runs the transaction with the specified isolation level.
+func WithIsolationLevel(x sql.IsolationLevel) WithTxOption {
+	return func(c *withTxConfig) {
+		c.TxOptions.Isolation = x
+	}
+}
+
+// WithReadOnly causes WithTx to run the transaction as read only
+func WithReadOnly() WithTxOption {
+	return func(c *withTxConfig) {
+		c.ReadOnly = true
+	}
+}
+
+// WithMaxRetries sets the number of times to retry a transaction because of a serialization failure.
+func WithMaxRetries(n int) WithTxOption {
+	return func(c *withTxConfig) {
+		c.MaxRetries = n
+	}
+}
+
+// WithTx calls cb with a transaction,
+// The transaction is committed IFF cb returns nil.
+// If cb returns an error the transaction is rolled back.
+func WithTx(ctx context.Context, db *sqlx.DB, cb func(tx *sqlx.Tx) error, opts ...WithTxOption) (retErr error) {
+	c := &withTxConfig{
+		TxOptions: sql.TxOptions{
+			Isolation: sql.LevelSerializable,
+		},
+		MaxRetries: 100,
+	}
+	for _, opt := range opts {
+		opt(c)
+	}
+	backoffStrategy := backoff.NewExponentialBackOff()
```

Comment:
Discussed some changes offline. Moving this to a `backoff.RetryUntilCancel` and removing the `MaxRetries` configuration.